iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/pipeconsensus/PipeConsensusAsyncConnector.java

```diff
@@ -0,0 +1,585 @@
+    isTabletBatchModeEnabled = false;
+  }
+
+  /** Add an event to transferBuffer, whose events will be asynchronizedly transfer to receiver. */
+  private boolean addEvent2Buffer(Event event) {
+    try {
+      LOGGER.info(
+          "PipeConsensus-ConsensusGroup-{}: no.{} event-{} added to connector buffer",
+          consensusGroupId,
+          ((EnrichedEvent) event).getCommitId(),
+          event);
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(
+            "PipeConsensus connector: one event enqueue, queue size = {}, limit size = {}",
+            transferBuffer.size(),
+            COMMON_CONFIG.getPipeConsensusEventBufferSize());
+      }
+      boolean result =
+          transferBuffer.offer(
+              event,
+              COMMON_CONFIG.getPipeConsensusEventEnqueueTimeoutInMs(),
+              TimeUnit.MILLISECONDS);
+      // add reference
+      if (result) {
+        ((EnrichedEvent) event).increaseReferenceCount(PipeConsensusAsyncConnector.class.getName());
+      }
+      return result;
+    } catch (InterruptedException e) {
+      LOGGER.info("PipeConsensusConnector transferBuffer queue offer is interrupted.", e);
+      Thread.currentThread().interrupt();
+      return false;
+    }
+  }
+
+  /**
+   * if one event is successfully processed by receiver in PipeConsensus, we will remove this event
+   * from transferBuffer in order to transfer other event.
+   */
+  public synchronized void removeEventFromBuffer(Event event) {
+    LOGGER.info(
+        "PipeConsensus-ConsensusGroup-{}: one event-{} successfully received by the follower, will be removed from queue, queue size = {}, limit size = {}",
+        consensusGroupId,
+        event,
+        transferBuffer.size(),
+        COMMON_CONFIG.getPipeConsensusEventBufferSize());
+
+    Iterator<Event> iterator = transferBuffer.iterator();
+    Event current = iterator.next();
+    while (!current.equals(event) && iterator.hasNext()) {
+      current = iterator.next();
+    }
+    iterator.remove();
+    // decrease reference count
+    ((EnrichedEvent) event)
+        .decreaseReferenceCount(PipeConsensusAsyncConnector.class.getName(), true);
+    // decrease alreadySentEventsCounts
+    alreadySentEventsInTransferBuffer.decrementAndGet();
+  }
+
+  @Override
+  public void handshake() throws Exception {
+    // do nothing
+    // PipeConsensus doesn't need to do handshake, since nodes in same consensusGroup/cluster
+    // usually have same configuration.
+  }
+
+  @Override
+  public void heartbeat() throws Exception {
+    // do nothing
+  }
+
+  @Override
+  public void transfer(TabletInsertionEvent tabletInsertionEvent) throws Exception {
+    boolean enqueueResult = addEvent2Buffer(tabletInsertionEvent);
+    if (!enqueueResult) {
+      throw new PipeException(ENQUEUE_EXCEPTION_MSG);
+    }
+
+    syncTransferQueuedEventsIfNecessary();
+
+    if (!(tabletInsertionEvent instanceof PipeInsertNodeTabletInsertionEvent)
```

Comment:
"Enqueue" after the judgement, can be after "increaseReferenceCount".

Reply:
Good catch. fixed